dbms/src/Flash/Coprocessor/AggregationInterpreterHelper.cpp

```diff
@@ -76,34 +76,63 @@ bool isGroupByCollationSensitive(const Context & context)
     return context.getSettingsRef().group_by_collation_sensitive || context.getDAGContext()->isMPPTask();
 }
 
-Aggregator::Params buildParams(
+std::shared_ptr<Aggregator::Params> buildParams(
     const Context & context,
     const Block & before_agg_header,
     size_t before_agg_streams_size,
     size_t agg_streams_size,
     const Names & key_names,
+    const KeyRefAggFuncMap & key_ref_agg_func,
+    const AggFuncRefKeyMap & agg_func_ref_key,
     const TiDB::TiDBCollators & collators,
     const AggregateDescriptions & aggregate_descriptions,
     bool is_final_agg,
     const SpillConfig & spill_config)
 {
-    ColumnNumbers keys;
-    for (const auto & name : key_names)
+    ColumnNumbers keys(key_names.size(), 0);
+    size_t normal_key_idx = 0;
+    size_t agg_func_as_key_idx = key_names.size() - key_ref_agg_func.size();
+    // Put group by key that reference aggregate func after original key.
+    // For example: select sum(c0), first_row(c1), first_row(c3) group by c1, c2, c3
+    // Before: keys: c1 | c2 | c3
+    // After:  keys: c2 | c1 | c3
+    // By doing this, when deserialize group by keys from HashMap to columns,
+    // we only need to handle c2(convert_key_size == 1) and ignore c1/c3.
```

Comment:
Update comment to clarify that still need to let c1/c3 reference first_row agg func.